subsys/power/device_pm.c

```diff
@@ -0,0 +1,204 @@
+#include <logging/log.h>
+LOG_MODULE_DECLARE(power);
+
+/* Device PM request type */
+#define DEVICE_PM_ASYNC			(1 << 0)
+
+static u32_t pm_to_fsm_state(u32_t pm_state)
+{
+	if (pm_state == DEVICE_PM_ACTIVE_STATE) {
+		return DEVICE_PM_FSM_ACTIVE;
+	} else {
+		return DEVICE_PM_FSM_STATE_SUSPENDED;
+	}
+}
+
+static void device_pm_callback(struct device *dev,
+			       int retval, void *context, void *arg)
+{
+	u32_t target_state = *((u32_t *)context);
+	u32_t fsm_state = *((u32_t *)arg);
+
+	k_sem_take(&dev->pm->lock, K_FOREVER);
+	if ((fsm_state == dev->pm->fsm_state) &&
+		(target_state == dev->pm->target_state) && !retval) {
+		dev->pm->fsm_state = pm_to_fsm_state(target_state);
+		/* Notify the driver through signal */
+		k_poll_signal_raise(&dev->pm->signal, target_state);
+	} else {
+		k_work_submit(&dev->pm->work);
+	}
+	k_sem_give(&dev->pm->lock);
+}
+
+static void pm_work_handler(struct k_work *work)
+{
+	struct device_pm *pm = CONTAINER_OF(work,
+					struct device_pm, work);
+	struct device *dev = pm->dev;
+	int ret;
+	u32_t target_state, fsm_state;
+
+	k_sem_take(&dev->pm->lock, K_FOREVER);
+	target_state = dev->pm->target_state;
+	if (target_state == DEVICE_PM_ACTIVE_STATE) {
+		dev->pm->fsm_state = DEVICE_PM_FSM_STATE_RESUMING;
+	} else {
+		dev->pm->fsm_state = DEVICE_PM_FSM_STATE_SUSPENDING;
+	}
+	fsm_state = dev->pm->fsm_state;
+	k_sem_give(&dev->pm->lock);
+
+	ret = device_set_power_state(dev, target_state,
+				     device_pm_callback, &fsm_state);
+	if (ret < 0) {
+		LOG_ERR("Set power state(%d) error(%d)\n", target_state, ret);
+	}
+}
+
+static int device_pm_request(struct device *dev,
+			     u32_t target_state, u32_t pm_flags)
+{
+	int ret = 0;
+
+	__ASSERT((target_state == DEVICE_PM_ACTIVE_STATE) ||
+			(target_state == DEVICE_PM_SUSPEND_STATE),
+			"Invalid device PM state requested");
+
+	if (target_state == DEVICE_PM_ACTIVE_STATE) {
+		if (atomic_inc(&dev->pm->usage) < 0) {
+			goto req_out;
+		}
+	} else {
+		if (atomic_dec(&dev->pm->usage) > 1) {
+			goto req_out;
+		}
+	}
+
+	k_sem_take(&dev->pm->lock, K_FOREVER);
+	if (!dev->pm->enable) {
+		k_sem_give(&dev->pm->lock);
+		goto req_out;
+	}
+
+	if (dev->pm->target_state == target_state) {
+		if ((pm_flags & DEVICE_PM_ASYNC) &&
+			(pm_to_fsm_state(target_state) ==
+				dev->pm->fsm_state)) {
+			/* Notify the driver through signal */
+			k_poll_signal_raise(&dev->pm->signal,
+					dev->pm->target_state);
+		}
+		k_sem_give(&dev->pm->lock);
+		goto req_out;
+	}
+
+	dev->pm->target_state = target_state;
+	k_work_submit(&dev->pm->work);
+
+	k_sem_give(&dev->pm->lock);
+
+	/* Incase of Sync request wait for completion event */
+	if (!(pm_flags & DEVICE_PM_ASYNC)) {
+		u32_t signaled;
+		int result;
+
+		ret = k_poll(&dev->pm->event, 1, K_FOREVER);
+		if (ret < 0 || dev->pm->event.state != K_POLL_STATE_SIGNALED) {
+			ret = -EIO;
+		} else {
+			k_poll_signal_check(&dev->pm->signal,
+						&signaled, &result);
+			if (result != target_state) {
```

Comment:
ok will check.

Reply:
Still, you are waiting here for first signal, which might be not the you are waiting for.
Also, the signal state is not reset, so this will work only once.